source/MaterialXContrib/MaterialXNode/MaterialXData.cpp

```diff
@@ -1,49 +1,102 @@
-#include <MaterialXData.h>
-#include <Util.h>
+#include "MaterialXData.h"
+#include "MaterialXUtil.h"
 
 #include <MaterialXFormat/XmlIo.h>
 #include <MaterialXGenOgsXml/GlslFragmentGenerator.h>
 #include <MaterialXGenShader/Util.h>
 
-MaterialXData::MaterialXData(const std::string& materialXDocumentPath, 
-                             const std::string& elementPath, 
-                             const MaterialX::FileSearchPath& librarySearchPath)
-    : _shaderGenerator(mx::GlslFragmentGenerator::create()),
-      _genContext(_shaderGenerator),
-      _librarySearchPath(librarySearchPath)
+namespace
 {
-    setRenderableElement(materialXDocumentPath, elementPath);
+mx::DocumentPtr createDocument( const std::string& materialXDocumentPath,
+                                const MaterialX::FileSearchPath& librarySearchPath )
+{
+    // Create document
+    mx::DocumentPtr document = mx::createDocument();
+    if (!document)
+    {
+        throw mx::Exception("Failed to create a MaterialX document");
+    }
+
+    // Load libraries
+    static const mx::StringVec libraries = { "stdlib", "pbrlib", "bxdf", "stdlib/genglsl", "pbrlib/genglsl" };
+    MaterialXMaya::loadLibraries(libraries, librarySearchPath, document);
+
+    // Read document contents from disk
+    mx::XmlReadOptions readOptions;
+    readOptions.skipDuplicateElements = true;
+    mx::readFromXmlFile(document, materialXDocumentPath, mx::EMPTY_STRING, &readOptions);
+
+    return document;
+}
 }
 
-bool MaterialXData::setRenderableElement(const std::string& materialXDocument, const std::string& elementPath)
+MaterialXData::MaterialXData(   mx::DocumentPtr document,
+                                const std::string& elementPath, 
+                                const MaterialX::FileSearchPath& librarySearchPath )
+    : _document(document)
+    , _shaderGenerator(mx::GlslFragmentGenerator::create())
```

Comment:
The generator here is currently hard-coded to GLSL, you probably want to dynamically create the correct one based on a "language" and "target" string passed in. Libraries can be loaded dynamically here as well. Might be good to do it during the generation call. I know Maya cannot handle different devices but might be good to cleanup to handle DX11 which I assume will need to come.

Reply:
Makes sense, but isn't this future work for sometime after our demo? Before we have at least one functional alternative target like OGS + HLSL, any design we put in for this would not be testable.